pkg/controllers/node/termination/controller.go

```diff
@@ -107,6 +108,15 @@ func (c *Controller) finalize(ctx context.Context, node *v1.Node) (reconcile.Res
 		}
 		return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
 	}
+	// In order for stateful pods to smoothly migrate from the terminating Node, we wait for VolumeAttachments
+	// of drain-able pods to be cleaned up before terminating the node and removing it from the cluster.
+	areVolumesDetached, err := c.ensureVolumesDetached(ctx, node)
```

Comment:
Now that Karpenter supports TerminationGracePeriod (#916), we'll need to support a mechanism that ensures we don't stall here indefinitely if detachments fail. This would only kick in if TGP is configured, otherwise we would wait indefinitely. I don't think this should be too complicated, really we just need to check if currentTime - deletionTimestamp > TGP and proceed if it is.

Reply:
That's a good point, we don't currently differentiate between critical and non-critical pods when deleting for TGP. Not something I think we need to consider for this PR, but I think you're right that that would be the more correct behavior.